File: ruby/ONP.rb
class Node
	attr_accessor :Left, :Right, :self
end

def check( str, st, fi ) 
    for  i in st..fi
        if (str[i] == '/' || str[i] == '*' || str[i] == '+' || str[i] == '-' || str[i] == '^')
            return false;
        end
    end
    return true
end

def makeTree( str, st, fi)
	puts st.to_s + ' ' + fi.to_s

 	root = Node.new
	if ( check( str, st, fi) )
		if (str[st.to_i].to_s != '(')
			root.self = str[st.to_i]
		elsif 
			root.self = str[fi.to_i]
		end

		root.Left = root.Right = nil
	end

	k = 1
	minP = 2147483646
	minPos = -1

	for i in st..fi
		nowP = 2147483647
		
		if ((str[i]).to_s == '(')
			k += 3
		end

		if ((str[i]).to_s == ')')
			k -= 3
		end

		if ((str[i]).to_s == '+' || (str[i]).to_s == '-')
			minP = k
		end

		if ((str[i]).to_s == '*' || (str[i]).to_s == '/')
			minP = k + 1
		end

		if ((str[i]).to_s == '^')
			minP = k + 2
		end

		if nowP <= minP
			minP = nowP
			minPos = i
		end
	end

	root.self = str[minPos];
	root.Left = makeTree( str, st, minPos - 1);
	root.Right = makeTree( str, minPos + 1, fi);
	return root
end

def dfs( root )
	if root.Left != nil 
		dfs(root.Left);
	end

	if root.Right 
		dfs(root.Right);
	end

	print root.self;
end
	
t = gets
t = t.to_i
while (t > 0)
	str = gets
	root = makeTree(str, 0, str.length-2);
	dfs( root );
	t = t - 1;
end

File: cpp/ONP.cpp
#include <iostream>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>

using namespace std;

struct node
{
	node* L;
	node* R;
	char self;
};

bool check(string str, int st, int fi)
{
    for (int i=st; i<=fi; i++)
        if (str[i] == '/' || str[i] == '*' || str[i] == '+' || str[i] == '-' || str[i] == '^')
            return false;
    return true;
}

node* makeTree( string str, int st, int fi )
{
	node* root = new node;
	if (check(str, st, fi))
	{
	    if (str[st] != '(')
            root->self = str[st];
        else
            root->self = str[fi];

		root->L = root->R = NULL;
		return root;
	}

	int k = 1;
	int minP = 2147483646;
	int minPos = -1;

	for (int i = st; i <= fi; ++i)
	{
		int nowP = 2147483647;
		if (str[i] == '(') k += 3;
		if (str[i] == ')') k -= 3;
		if (str[i] == '+') nowP = k;
		if (str[i] == '-') nowP = k;
		if (str[i] == '*') nowP = k + 1;
		if (str[i] == '/') nowP = k + 1;
		if (str[i] == '^') nowP = k + 2;

		if (nowP <= minP)
		{
			minP = nowP;
			minPos = i;
		}
	}

	root->self = str[minPos];

	root->L = makeTree( str, st, minPos-1);
	root->R = makeTree( str, minPos+1, fi);

	return root;
}

void dfs( node* root)
{
	if (root->L) dfs(root->L);
	if (root->R) dfs(root->R);
	printf("%c", root->self);
}

int main(int argc, char const *argv[])
{
	int t; scanf("%d", &t);
	while (t--)
	{
		string str; cin >> str;
		node *root;
		root = makeTree(str, 0, str.length()-1);
		dfs(root);
		printf("\n");
	}
	return 0;
}
